lazy_dataset/core.py

```diff
@@ -2313,6 +2316,106 @@ def __iter__(self):
             print(f'Dropped {dropped_count} examples')
 
 
+class ProfilingDataset(Dataset):
+    """
+    Special dataset to profile each dataset.
+    The repr will be modified to display the time that was spend in a
+    particular part of the pipeline.
+    
+    Warning:
+        This Dataset has limited support for prefetching. Only GIL bounded
+        prefetching is supported (e.g. thread pool and not process pool).
+
+    Example:
+        >>> import lazy_dataset
+        >>> ds = lazy_dataset.new({'a': 1, 'b': 2, 'c': 3})
+        >>> def sleep(ex):
+        ...     time.sleep(1)
+        ...     return ex
+        >>> ds_copy = ProfilingDataset(ds.map(sleep))
+        >>> _ = list(ds_copy)
+        >>> print(repr(ds_copy))  # doctest: +ELLIPSIS
+            DictDataset(len=3) (time = 0:00:00.0000...)
+          MapDataset(_pickle.loads) (time = 0:00:00.000...)
+        MapDataset(<function sleep at 0x...>) (time = 0:00:03.00...)
+        >>> ds_copy = ProfilingDataset(ds.map(sleep).prefetch(4, 8))
+        >>> _ = list(ds_copy)
+        >>> print(repr(ds_copy))  # doctest: +ELLIPSIS
+              DictDataset(len=3) (time = 0:00:00.0000...)
+            MapDataset(_pickle.loads) (time = 0:00:00.000...)
+          MapDataset(<function sleep at 0x...>) (time = 0:00:03.00...)
+        PrefetchDataset(4, 8, 't') (time = 0:00:01.0...)
+
+    """
+    # alternative time.process_time
+    timestamp = staticmethod(time.perf_counter)
+
+    def __init__(self, input_dataset):
+        if isinstance(input_dataset, self.__class__):
+            # This is not necessary, but the output is difficult to interpret.
+            raise RuntimeError(
+                'You can use ProfilingDataset only once.\n'
+                'The input_dataset is already a ProfilingDataset:\n'
+                f'{repr(input_dataset)}'
+            )
+        input_dataset = input_dataset.copy()
+
+        # use list with one element as mutable container to share the timer
+        # between the copies (Necessary for prefetch)
+        self.time = [0.0]
+
+        if hasattr(input_dataset, 'input_datasets'):
+            input_dataset.input_datasets = [
+                self.__class__(ds)
+                for ds in input_dataset.input_datasets
+            ]
+        if hasattr(input_dataset, 'input_dataset'):
+            input_dataset.input_dataset = ProfilingDataset(
+                input_dataset.input_dataset)
+
+        self.input_dataset = input_dataset
+
+    def __repr__(self):
+        r = repr(self.input_dataset)
+        # Better alternative for the name "time"?
+        r += f' (time = {datetime.timedelta(seconds=self.time[0])})'
```

Comment:
How about fetch_duration?

Reply:
Ok, I will rename `time` to `fetch duration`.